spec.md

```diff
@@ -79,15 +79,256 @@ Registry providers can self-certify by submitting conformance results to [openco
+```
+```
+<upload byte stream of chunk>
+```
+
+The `<location>` refers to the URL obtained from the preceding `POST` request.
+
+The `<range>` refers to the byte range of the chunk, and MUST be inclusive on both ends.  The first chunk's range MUST begin with `0`. It MUST match the following regular expression:
+
+```regexp
+^[0-9]+-[0-9]+$
+```
+
+The `<length>` is the content-length, in bytes, of the current chunk.
+
+Each successful chunk upload MUST have a `202 Accepted` response code, and MUST have the following header:
+
+```
+Location <location>
+```
+
+Each consecutive chunk upload SHOULD use the `<location>` provided in the response to the previous chunk upload.
+
+Chunks MUST be uploaded in order, with the first byte of a chunk being the last chunk's `<end-of-range>` plus one. If a chunk is uploaded out of order, the registry MUST respond with a `416 Requested Range Not Satisfiable` code.
+
+The final chunk MAY be uploaded using a `PATCH` request or it MAY be uploaded in the closing `PUT` request. Regardless of how the final chunk is uploaded, the session MUST be closed with a `PUT` request.
+
+---
+
+To close the session, issue a `PUT` request to a url in the following format, and with the following headers (and optional body, depending on whether or not the final chunk was uploaded already via a `PATCH` request):
+
+`<location>?digest=<digest>`
+```
+Content-Length: <length of chunk, if present>
+Content-Range: <range of chunk, if present>
+Content-Type: application/octet-stream <if chunk provided>
+```
+```
+OPTIONAL: <final chunk byte stream>
+```
+
+The closing `PUT` request MUST include the `<digest>` of the whole blob (not the final chunk) as a query parameter.
+
+The response to a successful closing of the session MUST be `201 Created`, and MUST contain the following header:
+```
+Location: <blob-location>
+```
+
+Here, `<blob-location>` is a pullable blob URL.
+
+
+##### Pushing Manifests
+
+To push a manifest, perform a `PUT` request to a path in the following format, and with the following headers
+and body:
+`/v2/<name>/manifests/<reference>`
+```http request
+Content-Type: application/vnd.oci.image.manifest.v1+json
+```
+```
+<manifest byte stream>
+```
+
+`<name>` is the namespace of the repository, and the `<reference>` MUST be either a) a digest or b) a tag.
+```regexp
+^[0-9]+-[0-9]+$
+```
+
+The uploaded manifest MUST reference any layers that make up the repository. However, the layers field MAY
+be empty. Upon a successful upload, the registry MUST return response code `201 Created`, and MUST have the
+following header:
+
+```http request
+Location: <location>
+```
+
+The `<location>` is a pullable manifest URL.
+
+An attempt to pull a nonexistent repository MUST return response code `404 Not Found`
 
 #### Content Discovery
 
-TODO: describe the Content Discovery category and the high-level details
+Content Discovery provides a way of discovering which tags are available for a given repository.
```

Comment:
The category should encapsulate more. Maybe make more general, and say "Currently, the only functionality provided by this workflow is the ability to discover tags ..."